### PR TITLE
Add relative layout to calendar

### PR DESCRIPTION
**Pull request by duanxianbo**

### CODE REVIEW

File: src/components/internal/FilterList.component.tsx [hunk 1]
```diff
@@ -71,17 +71,21 @@ export class FilterListComponent extends Component<Props, State> {
                     <div className={`${styles.Input} ${styles.TextField}`}>
                         <DateTextPickerUI
                             ref={this.startDataPickerRef}
+                            relativeLayout={this.props.deps.relativeLayout}
                             value={this.props.taskFiltersData.startDate}
                             label={'Start'}
                             onChange={this.onStartDateChange}
+                            zOffset={101}
```
Comment (naturetime10) [inline src/components/internal/FilterList.component.tsx]: Can we move this constant to the top of this file and give it a meaningful name?

File: src/components/internal/FilterList.component.tsx [hunk 2]
```diff
@@ -71,17 +71,21 @@ export class FilterListComponent extends Component<Props, State> {
...
                         />
                     </div>
                     <div className={`${styles.Input} ${styles.TextField}`}>
                         <DateTextPickerUI
                             ref={this.endDataPickerRef}
+                            relativeLayout={this.props.deps.relativeLayout}
                             value={this.props.taskFiltersData.endDate}
                             label={'End'}
                             onChange={this.onEndDateChange}
+                            zOffset={101}
```
Comment (naturetime10) [inline src/components/internal/FilterList.component.tsx]: Can we move this constant to the top of this file and give it a meaningful name?

File: src/components/internal/Modals/CreateSprintModal.component.tsx [hunk 3]
```diff
@@ -48,9 +48,11 @@ export class CreateSprintModalComponent extends Component<Props> {
                 <div className={styles.Content}>
                     <div className={styles.StartDate}>
                         <DateTextPickerUI
+                            relativeLayout={this.props.deps.relativeLayout}
                             label={'Start Date'}
                             ref={this.startDateInputRef}
                             value={new Date()}
+                            zOffset={1002}
```
Comment (naturetime10) [inline src/components/internal/Modals/CreateSprintModal.component.tsx]: Can we move this constant to the top of this file and give it a meaningful name?

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 4]
```diff
@@ -197,9 +197,11 @@ export class TaskDetailModalComponent extends Component<Props, State> {
                                 <div className={styles.Attributes}>
                                     <div className={styles.Row}>
                                         <DateTextPickerUI
+                                            relativeLayout={this.props.deps.relativeLayout}
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx:202]: Can we move this constant to the top of this file and give it a meaningful name?

File: src/libs/ui/DatePickerContainer.tsx [hunk 5]
```diff
@@ -1,7 +1,13 @@
+
 import moment from 'moment';
 import React, { Component, ReactNode } from 'react';
 
+import { RelativeLayout } from '@lib/layout/relativeLayout';
+
 import { today } from '../../components/internal/format';
+import {
+    PopupContainerUI,
+} from '../../components/internal/PopupContainer';
```
Comment (naturetime10) [inline src/libs/ui/DatePickerContainer.tsx]: Libs should only reference external package or other libs. Can we move `PopupContainer` into @libs/ui ?

File: src/components/internal/DurationInput.ui.tsx [hunk 6]
```diff
@@ -21,47 +20,37 @@ interface Props {
...
-                    {duration && (
-                        <div className={styles.Value}>
-                            {this.renderDuration()}
+                <PopupContainerUI
+                    overSpaceAction="jumpVertically"
+                    onOpen={this.props.onPopupOpen}
+                    onClose={this.props.onPopupClose}
+                    relativeLayout={this.props.relativeLayout}
+                    zOffset={this.props.zOffset}
+                    trigger={({ onToggle }) => (
```
Comment (naturetime10) [inline src/components/internal/DurationInput.ui.tsx]:
```suggestion
                    renderReferenceElement={({ onToggle }) => (
```

File: src/components/internal/DurationInput.ui.tsx [hunk 7]
```diff
@@ -21,47 +20,37 @@ interface Props {
...
-                    ref={this.popupRef}
-                    duration={duration}
-                    relativeLayout={this.props.relativeLayout}
-                    zOffset={(this.props.zOffset || 0) + 1}
-                    onDurationChange={this.props.onDurationChange}
-                    onMouseEnter={this.onPopupEnter}
-                    onMouseLeave={this.onPopupLeave}
-                    onOpen={this.props.onPopupOpen}
-                    onClose={this.onPopupClose}
+                    target={({ onClose }) => (
```
Comment (naturetime10) [inline src/components/internal/DurationInput.ui.tsx]:
```suggestion
                    renderPopup={({ onClose }) => (
```

File: src/components/internal/Modals/CreateSprintModal.component.tsx (hunk 3 above)
Comment (naturetime10) [inline src/components/internal/Modals/CreateSprintModal.component.tsx]: Can we move this constant to the top of this file and give it a meaningful name?

File: src/components/internal/FilterList.component.tsx (hunk 2 above)
Comment (naturetime10) [reply src/components/internal/FilterList.component.tsx]: Recommend naming: xxxxxZIndex
zOffset={xxxxxZIndex + 1}

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 8]
```diff
@@ -197,9 +197,11 @@ export class TaskDetailModalComponent extends Component<Props, State> {
...
                                             ref={this.dueAtDatePicker}
                                             label={'Due at'}
                                             value={dueAt}
+                                            zOffset={1002}
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]: Can we move this constant to the top of this file and give it a meaningful name?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/components/internal/FilterList.component.tsx (hunk 1 above)
Comment (duanxianbo) [reply src/components/internal/FilterList.component.tsx]: done

File: src/components/internal/FilterList.component.tsx (hunk 2 above)
Comment (duanxianbo) [reply src/components/internal/FilterList.component.tsx]: should we change it to snack_case as it is a constant?

File: src/components/internal/Modals/CreateSprintModal.component.tsx (hunk 3 above)
Comment (duanxianbo) [reply src/components/internal/Modals/CreateSprintModal.component.tsx]: done

File: src/components/internal/Modals/TaskDetailModal.component.tsx (hunk 4 above)
Comment (duanxianbo) [reply src/components/internal/Modals/TaskDetailModal.component.tsx:202]: done

File: src/libs/ui/DatePickerContainer.tsx (hunk 5 above)
Comment (duanxianbo) [reply src/libs/ui/DatePickerContainer.tsx]: changed name and moved to the ui/layout folder

File: src/components/internal/Modals/CreateSprintModal.component.tsx (hunk 3 above)
Comment (duanxianbo) [reply src/components/internal/Modals/CreateSprintModal.component.tsx]: done

File: src/components/internal/Modals/TaskDetailModal.component.tsx (hunk 8 above)
Comment (duanxianbo) [reply src/components/internal/Modals/TaskDetailModal.component.tsx]: done

File: src/components/internal/DurationInput.ui.tsx [hunk 9]
```diff
@@ -1,13 +1,13 @@
...
+import { RelativeLayout, RelativeLayoutContainerUI } from '@lib/layout';
 import { ButtonUI } from '@lib/ui/Button';
 import { SpacerUI } from '@lib/ui/Spacer';
 import { TextFieldUI } from '@lib/ui/TextField';
 
 import { Duration } from '@core/entity/duration';
 
 import styles from './DurationInput.ui.module.scss';
-import { PopupContainerUI } from './PopupContainer';
+
```
Comment (naturetime10) [inline src/components/internal/DurationInput.ui.tsx]:
```suggestion
```

File: src/components/internal/DurationInput.ui.tsx [hunk 10]
```diff
@@ -44,7 +44,7 @@ export class DurationInputUI extends Component<Props> {
                             )}
                         </div>
                     )}
-                    target={({ onClose }) => (
+                    renderPopup={({ onClose }) => (
```
Comment (naturetime10) [inline src/components/internal/DurationInput.ui.tsx]:
```suggestion
                    renderFollower={({ onClose }) => (
```

File: src/components/internal/FilterList.component.tsx [hunk 11]
```diff
@@ -75,7 +77,7 @@ export class FilterListComponent extends Component<Props, State> {
                             value={this.props.taskFiltersData.startDate}
                             label={'Start'}
                             onChange={this.onStartDateChange}
-                            zOffset={101}
+                            zOffset={DATE_PICKER_ZINDEX}
```
Comment (naturetime10) [inline src/components/internal/FilterList.component.tsx]:
```suggestion
                            zOffset={DATE_PICKER_Z_INDEX}
```

File: src/components/internal/FilterList.component.tsx [hunk 12]
```diff
@@ -85,7 +87,7 @@ export class FilterListComponent extends Component<Props, State> {
                             value={this.props.taskFiltersData.endDate}
                             label={'End'}
                             onChange={this.onEndDateChange}
-                            zOffset={101}
+                            zOffset={DATE_PICKER_ZINDEX}
```
Comment (naturetime10) [inline src/components/internal/FilterList.component.tsx]:
```suggestion
                            zOffset={DATE_PICKER_Z_INDEX}
```

File: src/components/internal/Modals/CreateSprintModal.component.tsx [hunk 13]
```diff
@@ -15,6 +15,7 @@ import { DURATIONS, DurationKey, getEndAt } from '../format';
 import styles from './CreateSprintModal.component.module.scss';
 
 const SPRINT_LENGTHS: DurationKey[] = ['1week', '2weeks'];
+const DATE_PICKER_ZINDEX = 1002
```
Comment (naturetime10) [inline src/components/internal/Modals/CreateSprintModal.component.tsx]:
```suggestion
const DATE_PICKER_Z_INDEX = 1002;
```

File: src/components/internal/Modals/CreateSprintModal.component.tsx [hunk 14]
```diff
@@ -52,7 +53,7 @@ export class CreateSprintModalComponent extends Component<Props> {
                             label={'Start Date'}
                             ref={this.startDateInputRef}
                             value={new Date()}
-                            zOffset={1002}
+                            zOffset={DATE_PICKER_ZINDEX}
```
Comment (naturetime10) [inline src/components/internal/Modals/CreateSprintModal.component.tsx]:
```suggestion
                            zOffset={DATE_PICKER_Z_INDEX}
```

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 15]
```diff
@@ -33,7 +33,8 @@ import { getShortName } from '../format';
 import styles from './TaskDetailModal.component.module.scss';
 
 const verticalMargin = 30;
-
+const DATE_PICKER_ZINDEX = 1002
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
const DATE_PICKER_Z_INDEX = 1002;
```

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 16]
```diff
@@ -33,7 +33,8 @@ import { getShortName } from '../format';
...
+const DURATION_INPUT_ZINDEX = 1000
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
const DURATION_INPUT_Z_INDEX = 1000;

```

File: src/libs/layout/RelativeLayoutContainer.tsx [hunk 17]
```diff
@@ -14,8 +14,8 @@ type Props = {
     onClose?: () => void;
     onMouseLeave?: () => void;
     zOffset?: number;
-    trigger: ({ onToggle }: { onToggle: () => void }) => ReactNode;
-    target: ({
+    renderReferenceElement: ({ onToggle }: { onToggle: () => void }) => ReactNode;
+    renderPopup: ({
```
Comment (naturetime10) [inline src/libs/layout/RelativeLayoutContainer.tsx]:
```suggestion
    renderFollower: ({
```

File: src/libs/layout/RelativeLayoutContainer.tsx [hunk 18]
```diff
@@ -33,7 +33,7 @@ interface States {
     showPopup: boolean;
 }
 
-export class PopupContainerUI extends Component<Props, States> {
+export class RelativeLayoutContainerUI extends Component<Props, States> {
     private readonly popupContainer = document.createElement('div');
     private readonly popupRef = createRef<HTMLDivElement>();
     private readonly triggerRef = createRef<HTMLDivElement>();
```
Comment (naturetime10) [inline src/libs/layout/RelativeLayoutContainer.tsx]: Discussed offline. Please change all `popup` keyword in this PR to `follower`

File: src/libs/ui/DatePickerContainer.tsx [hunk 19]
```diff
@@ -2,14 +2,12 @@
 import moment from 'moment';
 import React, { Component, ReactNode } from 'react';
 
-import { RelativeLayout } from '@lib/layout/relativeLayout';
-
+import { RelativeLayout, RelativeLayoutContainerUI } from '@lib/layout';
 import { today } from '../../components/internal/format';
```
Comment (naturetime10) [inline src/libs/ui/DatePickerContainer.tsx]: Can we only reference helpers from ui?

File: src/libs/ui/DatePickerContainer.tsx [hunk 20]
```diff
@@ -2,14 +2,12 @@
...
-import {
-    PopupContainerUI,
-} from '../../components/internal/PopupContainer';
+
 import { CalendarUI } from './Calendar';
 
+
```
Comment (naturetime10) [inline src/libs/ui/DatePickerContainer.tsx]:
```suggestion
```

File: tsconfig.json [hunk 21]
```diff
@@ -1,21 +1,27 @@
...
-        "noFallthroughCasesInSwitch": true,
-        "module": "esnext",
-        "moduleResolution": "node",
-        "resolveJsonModule": true,
-        "isolatedModules": true,
-        "noEmit": true,
-        "jsx": "react-jsx"
-    },
-    "include": ["src"]
+  "extends": "./tsconfig.path.json",
```
Comment (naturetime10) [inline tsconfig.json:2]: Let's add it to .gitignore

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/ui/DatePickerContainer.tsx (hunk 19 above)
Comment (duanxianbo) [reply src/libs/ui/DatePickerContainer.tsx]: done

File: tsconfig.json (hunk 21 above)
Comment (duanxianbo) [reply tsconfig.json:2]: done

File: src/libs/layout/RelativeLayoutContainer.tsx (hunk 18 above)
Comment (duanxianbo) [reply src/libs/layout/RelativeLayoutContainer.tsx]: done

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**